### PR TITLE
Fix metasm_shell by requiring readline

### LINKED ISSUE #17590: tools/exploit/metasm_shell is broken with uninitialized constant
```
tools/exploit/metasm_shell.rb -a Ia32
type "exit" or "quit" to quit
use ";" or "\n" for newline
type "file <file>" to parse a GAS assembler source file

/home/smcintyre/Repositories/metasploit-framework/lib/rex/ui/text/shell/history_manager.rb:63:in `clear_readline': uninitialized constant #<Class:#<HistoryManager stack size: 0>>::Readline (NameError)
	from /home/smcintyre/Repositories/metasploit-framework/lib/rex/ui/text/shell/history_manager.rb:99:in `switch_context'
	from /home/smcintyre/Repositories/metasploit-framework/lib/rex/ui/text/shell/history_manager.rb:29:in `push_context'
	from /home/smcintyre/Repositories/metasploit-framework/lib/rex/ui/text/shell.rb:133:in `run'
	from tools/exploit/metasm_shell.rb:171:in `<main>'
```

I'm pretty sure this has something to do with the new history manager and something needs to be loaded and initialized.

## Steps to reproduce

1. Run `tools/exploit/metasm_shell.rb`

### PR DESCRIPTION
**Pull request by zeroSteiner**
Fixes #17590 by adding a missing `require 'readline'` statement. Also fixes an exception due to the PseudoShell not providing `#run_single` that would occur when the shell was exited by the user pressing `Ctrl+D`.

## Verification

- [ ] Start `tools/exploit/metasm_shell.rb`
- [ ] Do not see a stack trace
- [ ] Exit the shell via `Ctrl+D` (EOF)
- [ ] Do not see a stack trace

## Before
```
tools/exploit/metasm_shell.rb       
type "exit" or "quit" to quit
use ";" or "\n" for newline
type "file <file>" to parse a GAS assembler source file

/home/smcintyre/Repositories/metasploit-framework/lib/rex/ui/text/shell/history_manager.rb:63:in `clear_readline': uninitialized constant #<Class:#<HistoryManager stack size: 0>>::Readline (NameError)
	from /home/smcintyre/Repositories/metasploit-framework/lib/rex/ui/text/shell/history_manager.rb:99:in `switch_context'
	from /home/smcintyre/Repositories/metasploit-framework/lib/rex/ui/text/shell/history_manager.rb:29:in `push_context'
	from /home/smcintyre/Repositories/metasploit-framework/lib/rex/ui/text/shell.rb:133:in `run'
	from tools/exploit/metasm_shell.rb:171:in `<main>'
```

## After
```
tools/exploit/metasm_shell.rb
type "exit" or "quit" to quit
use ";" or "\n" for newline
type "file <file>" to parse a GAS assembler source file

metasm > pop eax
"\x58"
metasm > exi
```

### CODE REVIEW

File: lib/rex/ui/text/shell/history_manager.rb [hunk 1]
```diff
@@ -1,4 +1,5 @@
 # -*- coding: binary -*-
+require 'readline'
```
Comment (adfoster-r7) [inline lib/rex/ui/text/shell/history_manager.rb:2]: I'm not sure if this is the right fix based on https://github.com/rapid7/metasploit-framework/blob/fd43c43a30e42e02058bd03455d37fff7e7342b0/lib/msf/ui/console/driver.rb#L658

Potentially the readline require belongs elsewhere, i.e. in the tool itself

File: lib/rex/ui/text/shell.rb [hunk 2]
```diff
@@ -149,7 +149,13 @@ def run(&block)
...
           self.stop_count += 1
           next if self.stop_count > 1
-          run_single('quit')
+
+          if block
+            block.call('quit')
+          elsif respond_to?(:run_single)
+            # PseudoShell does not provide run_single
+            run_single('quit')
+          end
```
Comment (gwillcox-r7) [inline lib/rex/ui/text/shell.rb:158]: Do we have a way to handle this if no block was passed and the code does not respond to `run_single`? Will this not end up falling through to the  following code which will also end up calling `run_single` without appropriate checks?

**Review by gwillcox-r7 — COMMENTED** [review]
One concern listed below, otherwise this looks good.

File: lib/rex/ui/text/shell.rb (hunk 2 above)
Comment (zeroSteiner) [reply lib/rex/ui/text/shell.rb:158]: Well if there's no block to run commands and there's no `#run_single` method to run commands, I'm not sure it's necessary or possible to run the `quit` command. In this case, falling through seems reasonable. If there's someplace where falling through without running the `quit` command seems wrong, let me know.

File: lib/rex/ui/text/shell/history_manager.rb (hunk 1 above)
Comment (zeroSteiner) [reply lib/rex/ui/text/shell/history_manager.rb:2]: Yeah that's probably a safer approach. I'll switch it over.

**Comment by zeroSteiner** [comment]
Apparently `nasm_shell` was affected by the same issue. I fixed that one too by adding the necessary require statement to the tool file.

**Review by gwillcox-r7 — APPROVED** [review]

**Comment by gwillcox-r7** [comment]
Before:

```
 ~/git/metasploit-framework │ master *1 ?20  ./tools/exploit/metasm_shell.rb                                                        ✔ │ 3s │ 3.0.5 Ruby │ 03:04:54 PM 
type "exit" or "quit" to quit
use ";" or "\n" for newline
type "file <file>" to parse a GAS assembler source file

/home/gwillcox/git/metasploit-framework/lib/rex/ui/text/shell/history_manager.rb:63:in `clear_readline': uninitialized constant #<Class:#<HistoryManager stack size: 0>>::Readline (NameError)
        from /home/gwillcox/git/metasploit-framework/lib/rex/ui/text/shell/history_manager.rb:99:in `switch_context'
        from /home/gwillcox/git/metasploit-framework/lib/rex/ui/text/shell/history_manager.rb:29:in `push_context'
        from /home/gwillcox/git/metasploit-framework/lib/rex/ui/text/shell.rb:133:in `run'
        from ./tools/exploit/metasm_shell.rb:171:in `<main>'
 ~/git/metasploit-framework │ master *1 ?20   
```

```
 ~/git/metasploit-framework │ master *1 ?20  ./tools/exploit/nasm_shell.rb                                                        1 ✘ │ 8s │ 3.0.5 Ruby │ 03:05:13 PM 
/home/gwillcox/git/metasploit-framework/lib/rex/ui/text/shell/history_manager.rb:63:in `clear_readline': uninitialized constant #<Class:#<HistoryManager stack size: 0>>::Readline (NameError)
        from /home/gwillcox/git/metasploit-framework/lib/rex/ui/text/shell/history_manager.rb:99:in `switch_context'
        from /home/gwillcox/git/metasploit-framework/lib/rex/ui/text/shell/history_manager.rb:29:in `push_context'
        from /home/gwillcox/git/metasploit-framework/lib/rex/ui/text/shell.rb:133:in `run'
        from ./tools/exploit/nasm_shell.rb:44:in `<main>'
 ~/git/metasploit-framework │ master *1 ?20      
```

**Comment by gwillcox-r7** [comment]
Seems to work fine, did have to rebase this against upstream for testing purposes but no issues encountered:

```
 ~/git/metasploit-framework │ land-pr17591:pr/17591 *1 ?20  git rebase -i upstream/master                                              1 ✘ │ 3.0.5 Ruby │ 03:07:19 PM 
Successfully rebased and updated refs/heads/land-pr17591.
 ~/git/metasploit-framework │ land-pr17591:pr/17591 ⇣2⇡108 *1 ?20  ./tools/exploit/metasm_shell.rb                                 ✔ │ 15s │ 3.0.5 Ruby │ 03:07:53 PM 
type "exit" or "quit" to quit
use ";" or "\n" for newline
type "file <file>" to parse a GAS assembler source file

metasm > %                                                                                                                                                               
 ~/git/metasploit-framework │ land-pr17591:pr/17591 ⇣2⇡108 *1 ?20  ./tools/exploit/nasm_shell.rb                                   ✔ │ 15s │ 3.0.5 Ruby │ 03:08:11 PM 
nasm > %                                                                                                                                                                 
 ~/git/metasploit-framework │ land-pr17591:pr/17591 ⇣2⇡108 *1 ?20          
```

**Comment by gwillcox-r7** [comment]
# Release Notes
A bug has been fixed in metasm_shell and nasm_shell whereby the shells were using readline but the dependency wasn't correctly imported. This has since been fixed and improved validation has been added.

**Pull request merged**